### DISCUSSION TITLE
Paint a sprite into a sprite with a viewport?

### DISCUSSION THREAD

**Discussion by chconnor** [Bodmer/TFT_eSPI#2967] [Q&A - Sprites] [answered]
Hi -- Let's say I want to paint a small sprite into bigger sprite at **x,y** in the big sprite's coordinate system, but I want to clip the small sprite at the boundary pixels of the bigger sprite. I have been doing this, and it works:

```
bigsprite.setViewport(1,1,WIDTH-2,HEIGHT-2);
littlesprite.pushToSprite(&bigsprite, x-1, y-1);
bigsprite.resetViewport();
```

...my question is about the **-1,-1** coordinates that seem necessary: is it normal/expected that the setViewport() changes the 0,0 location when calling pushToSprite? I would have expected to be able to do pushToSprite(&bigsprite, x, y) but that seems to draw offset by one pixel (or however many pixels I used in the coordinates passed to setViewport).

I just wanted to make sure that I'm not doing something that will cause a crash later by passing e.g. -1, -1 when the desired x,y is 0,0. :-)

Thanks!

**Comment by Bodmer** [comment]
It won't crash the processor, but it looks like a bug in the datum handling. I will push this into an Issue report until resolved.

**Comment by Bodmer** [comment]
I've created a simple sketch to test this:
```
#include <TFT_eSPI.h>
TFT_eSPI tft = TFT_eSPI();
TFT_eSprite spr1 = TFT_eSprite(&tft);  // Declare Sprite object "spr" with pointer to "tft" object
TFT_eSprite spr2 = TFT_eSprite(&tft);  // Declare Sprite object "spr" with pointer to "tft" object

void setup() {
  Serial.begin(115200);

  tft.init();

  tft.fillScreen(TFT_BLACK);

  spr1.createSprite(50, 50);
  spr1.fillSprite(TFT_WHITE);
  
  spr2.createSprite(50, 50);
  spr2.fillSprite(TFT_BLACK);
  spr2.fillRect(1,1,2,2,TFT_WHITE);
}

void loop()
{
  spr1.setViewport(1, 1, 48, 48);
  spr2.pushToSprite(&spr1,0,0);
  spr1.pushSprite(0,0);

  delay(2000);
}
```

It works fine on my setup. It should draw a white rectangle with a a 2x2 white pixel block at top left. This block should have a 1 pixel black margin at top and left.

Can you provide a simple sketch that demonstrates the problem you have noticed?

**Comment by chconnor** [comment]
Thanks! So, if that's the intended behavior, this may just be a misunderstanding on my part.

You call spr1.setViewport(1,1,48,48) and then spr2.pushToSprite(&spr1,0,0) -- for that pushToSprite call it appears that the coordinate system is now such that 0,0 is the top-left of the viewport, not the top-left of spr2, which is what surprised me (but perhaps should not have, I'm not experienced in these things.)

In other words, I interpreted setViewport to be setting a clipping mask, but that the coordinate system would still be 0,0 pointing to the top-left of the sprite, not the top-left of the viewport, which would make the small 2x2 block appear all the way in the corner, i.e. not with a 1-pixel black border around it.

So if your demo is showing the expected behavior (1-pixel black border around the 2x2 block) I will just keep doing the equivalent of -1,-1 to compensate for the viewport offset, since you said it will be harmless to do so?

**Reply by Bodmer** [reply, marked as answer]
Yes, the 0,0 datum by default moves to the corner of the viewport. This is [demonstrated here](https://github.com/Bodmer/TFT_eSPI/blob/master/examples/Generic/Viewport_Demo/Viewport_Demo.ino). Also you can optionally keep the datum at the top left corner of the screen or destination sprite. See [command parameters here](https://github.com/Bodmer/TFT_eSPI/blob/master/examples/Generic/Viewport_Demo/Viewport_commands.ino). So either keep using -1,-1 if you like, that is no problem, or add the extra true/false as needed to the function.
